### PR TITLE
Deprecated: Introduce new module with depreaction util

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Extracted from #6828.

Related issues: #3955, #6594.
Similar PRs: #6658, #6756, #6758.

This PR introduces new `@wordpress/deprecated` package. We might want to find a better name for it. I'm happy to update it if you have better ideas.

## How has this been tested?
`npm test`

Try `wp.deprecated( 'My message' )` in the JS console and make sure that deprecation warning is shown properly.

## Screenshots 
screen shot 2018-05-23 at 14 34 23

## Types of changes
Refactoring.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: packages/devtools/package.json
Comment (youknowriad) [inline packages/devtools/package.json]: I'm really bad at finding names so I'll leave this to others. 

Thinking that people could confuse `devtools` with actual dev tools not used in the code. Do you think we should try to find a name that indicates that it's a runtime dependency? I'm not certain there's a better name, just asking.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

File: packages/devtools/package.json
Comment (gziolo) [reply packages/devtools/package.json]: Me as well, I asked for help: https://wordpress.slack.com/archives/C5UNMSU4R/p1527080592000279.

**Comment by aduth** [comment]
Re: name, why not just `import deprecated from '@wordpress/deprecated';` ?

**Comment by gziolo** [comment]
> Re: name, why not just` import deprecated from '@wordpress/deprecated';`?

Done 👍

**Comment by gziolo** [comment]
Blocked by #6935 as we want to have this function exposed as `wp.deprecated`, not `wp.deprecated.default`. See #6748 for full context.

**Comment by gziolo** [comment]
This should be good now to merge.

File: .eslintrc.js
Comment (aduth) [inline .eslintrc.js]: Nit: This pattern should be `/^deprecated$/` (otherwise matching e.g. `deprecatedfoo`)

File: data/test/index.js
Comment (aduth) [inline data/test/index.js]: Would be neat to have something like our ESLint `deprecated` removal rules, except for things like this which aren't necessarily "deprecated", but are related to a deprecation or should otherwise be removed by a certain version. Maybe a comment pattern similar to [ESLint's inline configuration](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments), e.g. `/* REMOVEME: v3.0.0 */`.

File: utils/index.js
Comment (aduth) [inline utils/index.js]: Love the effort to rid the world of `./utils` bit by bit ❤️

File: .eslintrc.js
Comment (gziolo) [reply .eslintrc.js]: I noticed it too late, it is included together with changes planned for `data` package. Sorry about that :)

File: data/test/index.js
Comment (gziolo) [reply data/test/index.js]: Yes, it would help to remove it together with the code that was deprecated. Another option would be to include `toHaveWarned` in the code which would fail when deprecation gets removed. Unless I missed something.

**Pull request merged**
